File: apps/desktop/src/lib/tauri/cloak.ts
import type {
	CloakCapabilities,
	CloakInstallation,
	CloakValidationResult,
	DiscoveredCloakInstallation,
} from "@/types/cloak";

import { invokeCommand } from "./client";

export const cloakApi = {
	installation() {
		return invokeCommand<CloakInstallation>("cloak_get_installation");
	},

	setExecutable(path: string) {
		return invokeCommand<CloakInstallation>("cloak_set_executable", { path });
	},

	validate() {
		return invokeCommand<CloakValidationResult>("cloak_validate_installation");
	},

	capabilities() {
		return invokeCommand<CloakCapabilities>("cloak_get_capabilities");
	},

	discover() {
		return invokeCommand<DiscoveredCloakInstallation[]>("cloak_discover_installations");
	},

	autoConfigure() {
		return invokeCommand<CloakInstallation>("cloak_auto_configure");
	},
};

File: apps/desktop/src-tauri/src/commands/cloak.rs
use tauri::State;

use crate::application::services::CloakInstallationService;
use crate::domain::cloak::{
    CloakCapabilities, CloakInstallationDto, CloakValidationResult,
    DiscoveredCloakInstallationDto,
};
use crate::error::AppError;
use crate::state::AppState;

#[tauri::command]
pub async fn cloak_get_installation(
    state: State<'_, AppState>,
) -> Result<CloakInstallationDto, AppError> {
    CloakInstallationService::get_installation(&state).await
}

#[tauri::command]
pub async fn cloak_set_executable(
    state: State<'_, AppState>,
    path: String,
) -> Result<CloakInstallationDto, AppError> {
    CloakInstallationService::set_executable(&state, path).await
}

#[tauri::command]
pub async fn cloak_validate_installation(
    state: State<'_, AppState>,
) -> Result<CloakValidationResult, AppError> {
    CloakInstallationService::validate_installation(&state).await
}

#[tauri::command]
pub async fn cloak_discover_installations(
) -> Result<Vec<DiscoveredCloakInstallationDto>, AppError> {
    CloakInstallationService::discover_installations().await
}

#[tauri::command]
pub async fn cloak_auto_configure(
    state: State<'_, AppState>,
) -> Result<CloakInstallationDto, AppError> {
    CloakInstallationService::auto_configure(&state).await
}

#[tauri::command]
pub async fn cloak_get_capabilities(
    state: State<'_, AppState>,
) -> Result<CloakCapabilities, AppError> {
    CloakInstallationService::get_capabilities(&state).await
}
